pkg/bot/mattermost.go

```diff
@@ -109,24 +123,28 @@ func (b *MMBot) Start() error {
 		return fmt.Errorf("while pinging Mattermost server %q: %w", b.ServerURL, err)
 	}
 
-	go func() {
-		// It is observed that Mattermost server closes connections unexpectedly after some time.
-		// For now, we are adding retry logic to reconnect to the server
-		// https://github.com/infracloudio/botkube/issues/201
-		log.Info("BotKube connected to Mattermost!")
-		for {
+	// It is observed that Mattermost server closes connections unexpectedly after some time.
+	// For now, we are adding retry logic to reconnect to the server
+	// https://github.com/infracloudio/botkube/issues/201
+	b.log.Info("BotKube connected to Mattermost!")
+	for {
+		select {
+		case <-ctx.Done():
+			b.log.Info("Context canceled. Finishing...")
+			return nil
```

Comment:
Do we get error if a context is canceled? Should we check if the context has an error set?

Reply:
What do you mean by "Should we check if the context has an error set?"? Receiving a struct on `ctx.Done()` means the application needs to be terminated (SIGINT), so all components will return without an error. This is because in `main.go` we use this: https://pkg.go.dev/sigs.k8s.io/controller-runtime/pkg/manager/signals

You can try running BotKube from this PR and see how it shutdowns gracefully when you press CTRL+C (or CMD+C on Mac). Previously the app didn't react on the SIGINT signal and needed to be killed (SIGKILL) every time.

Reply:
In case the context gets canceled due to other reasons, should we check for [Err()](https://pkg.go.dev/context#Context).

Reply:
Oh, I get it now - but in this case it won't be useful:
```
	// If Done is closed, Err returns a non-nil error explaining why:
	// Canceled if the context was canceled
	// or DeadlineExceeded if the context's deadline passed.
```
As the context will always be `Canceled` and that's expected. I don't want to return error here as this is normal behavior and the app should exit with 0 after doing cleanup and sending final messages.

Reply:
IMO it brings consistency with other libs approach e.g.:
- gh cli: https://cs.github.com/cli/cli?q=ctx.Err%28%29
- or even Go itself: https://cs.github.com/golang/go?q=ctx.Err%28%29

I checked and only in one place we have additionally error, and using `multierror.Append` shouldn't be an issue there. However, if there will be a simple logic to return only the `dicsonnect` error, it would also work well.

---
> In most places, err handling in main it looks like this:

yep, but it can be handled with a simple func, which still makes it readable
```go
errGroup.Go(func() error {
  return logAndIgnoreCtxErr(cfgWatcher.Do(ctx, cancelCtxFn))
})
```


We also have the `errGroup` there, so it can be a single place for this err handling.

---

BTW, I also noticed that the `errGroup` is not the best approach for “graceful shutdown” handling, as it exits on the first non-nil error. If `ctx` is canceled, and `rtm.Disconnect()` returns an error, `Wait()` will be released, and we will exit without waiting for clean-up in other places 🤔 

With that being said, I'm 100% ok  with leaving it as it is right now 👍 If needed, we will address that later 🙂